primitives/src/v4/mod.rs

```diff
@@ -796,17 +799,93 @@ pub struct ParathreadEntry {
 	pub retries: u32,
 }
 
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Assignment {
+	pub para_id: Id,
+	pub collator_restrictions: CollatorRestrictions,
+}
+
+impl Assignment {
+	pub fn new(para_id: Id, collator_restrictions: CollatorRestrictions) -> Self {
+		Assignment { para_id, collator_restrictions }
+	}
+}
+
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct CollatorRestrictions {
+	/// Collators to prefer/allow.
+	/// Empty set means no restrictions.
+	collator_peer_ids: BTreeSet<OpaquePeerId>, // TODO: Better have CollatorIds here as we can have them in the runtime. And transform to PeerId on the node side?
```

Comment:
There is no way of transforming a `CollatorId` to a `PeerId`, except asking the connecting peer for its `CollatorId`, but that would completely defeat the point. Then we could equally well stick with the `CollatorId`.

Reply:
Yea, I realised as well there is no such transformation. Was just thinking that if there was, it would be nice to have something more strongly typed.

Reply:
Indeed. I brought this up a few times as well. Things would be easier if we would not have 10 different unrelated keys for the same thing :confused: Anyhow, in this particular case we can just go to the peer id directly and not bother about the collator id. It does not have any meaning on the relay chain anyway.